src/tools/immutable_object_cache/ObjectCacheStore.cc

```diff
@@ -0,0 +1,279 @@
+  ldout(m_cct, 20) << "object name = " << object_name
+                   << " in pool ID : " << pool_id << dendl;
+
+  int pret = -1;
+  std::string cache_file_name = std::move(get_cache_file_name(pool_nspace,
+                                            pool_id, snap_id, object_name));
+
+  cache_status_t ret = m_policy->lookup_object(cache_file_name);
+
+  switch(ret) {
+    case OBJ_CACHE_NONE: {
+      pret = do_promote(pool_nspace, pool_id, snap_id, object_name);
+      if (pret < 0) {
+        lderr(m_cct) << "fail to start promote" << dendl;
+      }
+      return -1;
+    }
+    case OBJ_CACHE_PROMOTED:
+      target_cache_file_path = std::move(
+        get_cache_file_path(cache_file_name));
+      return 0;
+    case OBJ_CACHE_SKIP:
+      return -1;
+    default:
+      lderr(m_cct) << "unrecognized object cache status." << dendl;
+      ceph_assert(0);
+  }
+}
+
+int ObjectCacheStore::promote_object(librados::IoCtx* ioctx,
+                                     std::string object_name,
+                                     librados::bufferlist* read_buf,
+                                     Context* on_finish) {
+  ldout(m_cct, 20) << "object name = " << object_name << dendl;
+
+  librados::AioCompletion* read_completion = create_rados_callback(on_finish);
+  // issue a zero-sized read req to get the entire obj
+  int ret = ioctx->aio_read(object_name, read_completion, read_buf, 0, 0);
+  if (ret < 0) {
+    lderr(m_cct) << "failed to read from rados" << dendl;
+  }
+  read_completion->release();
+
+  return ret;
+}
+
+int ObjectCacheStore::evict_objects() {
+  ldout(m_cct, 20) << dendl;
+
+  std::list<std::string> obj_list;
+  m_policy->get_evict_list(&obj_list);
+  for (auto& obj: obj_list) {
+    do_evict(obj);
+  }
+}
+
+int ObjectCacheStore::do_evict(std::string cache_file) {
+  ldout(m_cct, 20) << "file = " << cache_file << dendl;
+
+  if (cache_file == "") {
+    return 0;
+  }
+
+  std::string cache_file_path = std::move(get_cache_file_path(cache_file));
+
+  ldout(m_cct, 20) << "evict cache: " << cache_file_path << dendl;
+
+  // TODO(): possible race on read?
+  int ret = std::remove(cache_file_path.c_str());
+   // evict metadata
+  if (ret == 0) {
+    m_policy->update_status(cache_file, OBJ_CACHE_SKIP);
+    m_policy->evict_entry(cache_file);
+  }
+
+  return ret;
+}
+
+std::string ObjectCacheStore::get_cache_file_name(std::string pool_nspace,
+                                                       uint64_t pool_id,
+                                                       uint64_t snap_id,
+                                                       std::string oid) {
+  return pool_nspace + ":" + std::to_string(pool_id) + ":" +
+         std::to_string(snap_id) + ":" + oid;
+}
+
+std::string ObjectCacheStore::get_cache_file_path(std::string cache_file_name) {
+
+  std::string cache_file_dir = "";
+
+  if (m_dir_num > 0) {
+    auto const pos = cache_file_name.find_last_of('.');
```

Comment:
This needs to be generic -- seems to be expecting an object name with a period and base-10 numeric.